rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+```
+
+***future work***: At some point we may want to explore something like
+minimal invalidation for annotations; for a simple approach we could
+imagine not resending annotations of a given type when they are
+unchanged. For the time being, the simplicity of being entirely
+stateless is appealing.
+
+
+### In the frontend
+
+The annotation wire format is intended to be handled efficiently by
+frontends, particularly when drawing. Specifically, the annotations sent
+should include only those necessary to draw the current screen. within
+a type of annotations, items are in sorted, ascending order, meaning
+they can be processed in order, keeping a single index into both the
+offset and payload arrays.
+
+#### Annotations are stateless
+
+To the frontend, annotations are _stateless_. By this we mean that
+a given update will include the full set of annotations that intersect
+with the currently visible region; the client will not be expected to do
+any caching or merging. (As a possible optimization we might consider
+making this be true only _per type_; so that once received, annotations
+of a given type are considered valid until new annotations of that type
+are received.)
+
+#### Each type of annotation has a defined payload type.
+
+The payload associated with a given annotation is determined by the type
+of the annotation. We anticipate defining a set of types and their
+corresponding schemas as part of the xi-editor project, and more may be
+added in the future; but new types could also be defined by a plugin,
+and as long as they're supported by a frontend that's fine.
+
+Some examples of possible types and their possible payloads:
+
+**'`selection`'**: The payload for a selection is simply the location in
+the selection of the cursor. This might be represented by a number in
+`[0, 1, 2]`, corresponding to 'start', 'end/upstream', or
+'end/downstream'.
+
+**'`link`'** A link's payload might be a url string.
+
+**'`diagnostic`'** A diagnostic's payload might be an object including
+a level (e.g. warning, error, note), a short message string, and
+a longer detail string.
+
+
+#### The frontend chooses how to display annotations based on their type.
+
+The type of an annotation (and its documentation) should hint at how it
+might be displayed, but this choice is ultimately up to the frontend.
+Take for example a diagnostic warning annotation; a gui frontend might
+choose to show the warning inline, drawing over the text, but a tui
+frontend might only show an indicator in the gutter area:
+
+![vscode language server error](./assets/vscode_err.png)<br> _A language
+server error in Visual Studio Code_ <br>
+
+![neovim language server error](./assets/nvim_err.png)<br> _A language
+server error in Neovim_
+
+![Xcode error](./assets/xcode_err.png)<br> _An error in Xcode, just for
+variety_
+
+### plugin API
+
+Plugins add, update, and remove annotations with a single protocol
+method:
+
+```javascript
+{
+    "method": "update_annotations",
```

Comment:
This update needs to be based on a rev, as annotations will be subject to OT if the document has been edited concurrently with the annotation being in flight.